### PR TITLE
crud completed

### PR DESCRIPTION
**Pull request by kamal-web**

### CODE REVIEW

File: Server/app.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+var app = express();
+
+app.use(bodyParser.urlencoded({ extended: true }))
+app.use(bodyParser.json())
+
+mongoose.Promise = global.Promise;
+
+//Connecting to mongodb database
+function connect(){
+mongoose.connect("mongodb://localhost:27017/test",{ useNewUrlParser: true, useUnifiedTopology: true }).then(
```
Comment (yaswanthkoneri) [inline Server/app.js]: indentation is a bit off

File: Server/controllers/employeeCrud.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+//Create Emp
+ EmpController.prototype.create = (req,res) => {
+     if(!req.body){
+        return res.status(400).send({
+            message: "Details could not be empty"
+        });
+    }else{
+        const emp = new empDetails({
+            id        : req.body.id,
```
Comment (yaswanthkoneri) [inline Server/controllers/employeeCrud.js]: use object destructuring for getting content

`const {id, name and so on ..} = req.body`

File: Server/controllers/employeeCrud.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        .then(data => {
+            res.send(data);
+        }).catch(err => {
+            res.status(500).send({
+                message: err.message || "values are Not inserted"
+            })
+        })
+
+    }
+
```
Comment (yaswanthkoneri) [inline Server/controllers/employeeCrud.js]: remove all unwanted whitespaces

File: Server/controllers/employeeCrud.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+//Create Emp
+ EmpController.prototype.create = (req,res) => {
+     if(!req.body){
+        return res.status(400).send({
+            message: "Details could not be empty"
+        });
+    }else{
+        const emp = new empDetails({
```
Comment (yaswanthkoneri) [inline Server/controllers/employeeCrud.js]: rename empDetails to Employee

File: Server/controllers/employeeCrud.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
+const empDetails =require('../models/schema');
```
Comment (yaswanthkoneri) [inline Server/controllers/employeeCrud.js]: rename empDetails to Employee

File: Server/controllers/employeeCrud.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            message: err.message || "Details are not retrived"
+        })
+    })
+
+}
+
+//Find one Employee
+EmpController.prototype.findOne=(req,res)=>{
+    let{id} =req.params;
+    console.log(typeof(id))
```
Comment (yaswanthkoneri) [inline Server/controllers/employeeCrud.js]: remove consoles before raising PR
only logging consoles should be there

File: Server/controllers/employeeCrud.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
+const empDetails =require('../models/schema');
+
+
+function EmpController(){
+
+}
+
+//Create Emp
+ EmpController.prototype.create = (req,res) => {
+     if(!req.body){
```
Comment (yaswanthkoneri) [inline Server/controllers/employeeCrud.js]: Do validations, what if some data is missing?

File: Server/controllers/employeeCrud.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        })
+    })
+    
+
+}
+
+//Update Employee
+EmpController.prototype.update =(req, res)=> {
+    let{id} =req.params;
+    let emp = req.body;
```
Comment (yaswanthkoneri) [inline Server/controllers/employeeCrud.js]: dont use let, use const and object destructuring

File: Server/models/dbconfig.js [hunk 9]
```diff
@@ -0,0 +1,3 @@
+module.exports =  {
+    url: 'mongodb://localhost:27017/test'
```
Comment (yaswanthkoneri) [inline Server/models/dbconfig.js]: move this to config, and use config module 

`const config = require('config')`

File: Server/routes/emproutes.js [hunk 10]
```diff
@@ -0,0 +1,11 @@
+module.exports = (app) => {
```
Comment (yaswanthkoneri) [inline Server/routes/emproutes.js]: rename file name as employee.js

File: Server/.gitignore [hunk 11]
```diff
@@ -0,0 +1 @@
+node_modules/
```
Comment (yaswanthkoneri) [inline Server/.gitignore]: include config file also here

File: Server/controllers/employeeCrud.js [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        if(!emp){
+            res.status(500).send({
+                message:"Id not found, Please Enter valid Id"
+            })
+        }
+        else{
+            res.send(emp);
+        }
+    }).catch(err => {
+        res.status(500).send({
```
Comment (yaswanthkoneri) [inline Server/controllers/employeeCrud.js]: log errors in case of catch

console.error(err)

File: Server/routes/emproutes.js (hunk 10 above)
Comment (yaswanthkoneri) [inline Server/routes/emproutes.js]: Why do we need this file when we have routes/index.js
Please Explain? am I missing something?

File: Server/routes/index.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
+var express = require('express');
+var router = express.Router();
+var EmpController = require('../controllers/employeeCrud')
```
Comment (yaswanthkoneri) [inline Server/routes/index.js]: rename employeeCrud to employee.js

File: Server/routes/index.js [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+/* GET home page. */
+router.get('/', function(req, res, next) {
+  res.render('index', { title: 'Express' });
+});
+     // Create a new Note
+     router.post('/add', EmpController.create);
```
Comment (yaswanthkoneri) [inline Server/routes/index.js]: make this route to '/employee'

File: Server/routes/index.js [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+    // Retrieve all Notes
```
Comment (yaswanthkoneri) [inline Server/routes/index.js]: write appropriate comments

Comment (yaswanthkoneri) [reply Server/routes/index.js]: // Retrieve all Employee Details

File: Server/app.js (hunk 1 above)
Comment (yaswanthkoneri) [inline Server/app.js]: put this url into config file

Comment (yaswanthkoneri) [reply Server/app.js]: Can you write IIFE(Immediately invoked Function expression) for connect function

File: Server/controllers/employee.js [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+            })
+        }
+        else{
+            res.send(emp);
+        }
+    }).catch(err => {
+        res.status(500).send({
+            message: `Error:${err.message}`
+        })
+        console.log("values are Not inserted",err.message)
```
Comment (yaswanthkoneri) [inline Server/controllers/employee.js]: console before response not after response and for errors always use `console.error`

File: Server/controllers/employee.js [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+    })
+}
+
+//Update Employee
+EmpController.prototype.update =(req, res)=> {
+    const{checkId} =req.params;
+    const{id,name,email,department,role,salary} = req.body;
```
Comment (yaswanthkoneri) [inline Server/controllers/employee.js]: give space between const and flower bracket for better readability

File: Server/controllers/employee.js [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+    }).catch(err => {
+        res.status(500).send({
+            message: `Error:${err.message}`
+        })
+        console.log("values are Not inserted",err.message)
+    })
+}
+
+//Update Employee
+EmpController.prototype.update =(req, res)=> {
```
Comment (yaswanthkoneri) [inline Server/controllers/employee.js]: is your update code working? I suspect its not working

File: Server/controllers/employee.js [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+    const{checkId} =req.params;
+    const{id,name,email,department,role,salary} = req.body;
+    //console.log(emp.name)
+    Employee.findOneAndUpdate({checkId},{id:id,name:name,email:email,department:department,role:role,salary:salary}, {new: true})
```
Comment (yaswanthkoneri) [inline Server/controllers/employee.js]: Indentation is off

File: Server/controllers/employee.js [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+            })
+        }
+        else{
+            res.send(emp);
+        }
+    }).catch(err => {
+        res.status(500).send({
+            message: `Error:${err.message}`
+        })
+        console.log("values are Not inserted",err.message)
```
Comment (yaswanthkoneri) [inline Server/controllers/employee.js]: console.error before response

File: Server/controllers/employee.js [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+    })
+}
+
+//Delete Employee
+EmpController.prototype.delete = (req, res) => {
+    
+    let{id} =req.params;
```
Comment (yaswanthkoneri) [inline Server/controllers/employee.js]: use const instead of let

File: Server/controllers/employee.js [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+    console.log(id)
```
Comment (yaswanthkoneri) [inline Server/controllers/employee.js]: remove unwanted consoles

File: Server/controllers/employee.js [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+};
+
+//Create Emp
+ EmpController.prototype.create = (req,res) => {
+     if(!req.body){
+        return res.status(400).send({
+            message: "Details could not be empty"
+        });
+    }else{
+        if(!isValidArg(req.body.id)){
```
Comment (yaswanthkoneri) [inline Server/controllers/employee.js]: you could have destructured the object before and used it here as well

File: Server/controllers/employee.js [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+        if(!isValidArg(req.body.role)){
+            res.status(402).json({message:"please provide role"})
+        }
+        if(!isValidArg(req.body.salary)){
+            res.status(402).json({message:"please provide salary"})
+        }
+        if(!isValidArg(req.body.experience)){
+            res.status(402).json({message:"please provide experience"})
+        }
+        const{id,name,email,mobile,department,role,salary} = req.body
```
Comment (yaswanthkoneri) [inline Server/controllers/employee.js]: use space b/w const and flower braket always

File: Server/controllers/employee.js [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+        const emp = new Employee({id,name,email,mobile,department,role,salary});
+        emp.save()
+        .then(data => {
+            res.send(data);
+        }).catch(err => {
+            res.status(500).send({
+                message: err.message || "values are Not inserted"
+            })
+            console.log("values are Not inserted",err.message)
```
Comment (yaswanthkoneri) [inline Server/controllers/employee.js]: console before response

File: Server/app.js [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+mongoose.Promise = global.Promise;
+//Connecting to mongodb database
+function connect(){
+    mongoose.connect(dbConfig.url,{ useNewUrlParser: true, useUnifiedTopology: true })
+    .then(
+      () => { console.log("Connected to Mongodb Successfull") },
+      err => { return err }
+    );
+}
+connect()
```
Comment (yaswanthkoneri) [inline Server/app.js]: use IIFE for connect method

File: Server/routes/index.js [hunk 27]
```diff
@@ -0,0 +1,24 @@
+var express = require('express');
+var router = express.Router();
+var EmpController = require('../controllers/employee')
+
+/* GET home page. */
+router.get('/', function(req, res, next) {
```
Comment (yaswanthkoneri) [inline Server/routes/index.js]: remove this route, we dont need this anymore

File: Server/routes/emproutes.js (hunk 10 above)
Comment (kamal-web) [reply Server/routes/emproutes.js]: Its an old file which i have removed. I just took it for reference from my old project.

File: Server/app.js (hunk 1 above)
Comment (kamal-web) [reply Server/app.js]: done

File: Server/controllers/employeeCrud.js (hunk 2 above)
Comment (kamal-web) [reply Server/controllers/employeeCrud.js]: done

File: Server/controllers/employeeCrud.js (hunk 3 above)
Comment (kamal-web) [reply Server/controllers/employeeCrud.js]: done

File: Server/controllers/employeeCrud.js (hunk 4 above)
Comment (kamal-web) [reply Server/controllers/employeeCrud.js]: done

File: Server/controllers/employeeCrud.js (hunk 6 above)
Comment (kamal-web) [reply Server/controllers/employeeCrud.js]: done

File: Server/controllers/employeeCrud.js (hunk 7 above)
Comment (kamal-web) [reply Server/controllers/employeeCrud.js]: done

File: Server/controllers/employeeCrud.js (hunk 8 above)
Comment (kamal-web) [reply Server/controllers/employeeCrud.js]: done

File: Server/routes/index.js (hunk 13 above)
Comment (kamal-web) [reply Server/routes/index.js]: done

File: Server/routes/index.js (hunk 14 above)
Comment (kamal-web) [reply Server/routes/index.js]: done

File: Server/app.js (hunk 1 above)
Comment (kamal-web) [reply Server/app.js]: done

**Pull request merged**